devicetypes/smartthings/zigbee-battery-accessory-dimmer.src/zigbee-battery-accessory-dimmer.groovy

```diff
@@ -103,75 +149,149 @@ def handleSengledSwitchEvents(descMap) {
+			results << createEvent(zigbee.getEventFromAttrData(descMap.clusterInt, descMap.commandInt, UINT8_STR, descMap.data[0]))
 		}
 	}
+
+	return results
+}
+
+def handleSwitchEvent(descMap) {
+	def results = []
+
+	if (descMap.commandInt == ONOFF_ON_COMMAND) {
+		if (device.currentValue("level") == 0) {
+			results << createEvent(name: "level", value: DOUBLE_STEP)
+		}
+		results << createEvent(name: "switch", value: "on")
+	} else if (descMap.commandInt == ONOFF_OFF_COMMAND) {
+		results << createEvent(name: "switch", value: "off")
+	}
+
+	return results
+}
+
+def handleStepEvent(direction, descMap) {
+	def results = []
+	def currentLevel = device.currentValue("level") as Integer ?: 0
+	def value = null
+
+	if (direction == LEVEL_DIRECTION_UP) {
+		value = Math.min(currentLevel + DOUBLE_STEP, 100)
+	} else if (direction == LEVEL_DIRECTION_DOWN) {
+		value = Math.max(currentLevel - DOUBLE_STEP, 0)
+	}
+
+	if (value != null) {
+		log.debug "Step ${direction == LEVEL_DIRECTION_UP ? "up" : "down"} by $DOUBLE_STEP to $value"
+
+		// don't change level if switch will be turning off
+		if (value == 0) {
+			results << createEvent(name: "switch", value: "off")
+		} else {
+			results << createEvent(name: "switch", value: "on")
+			results << createEvent(name: "level", value: value)
+		}
+	} else {
+		log.debug "Received invalid direction ${direction} - descMap.data = ${descMap.data}"
+	}
+
+	return results
 }
 
 def handleBatteryEvents(descMap) {
-	def batteryValue = null
-	if (descMap?.clusterInt == zigbee.POWER_CONFIGURATION_CLUSTER && descMap?.value) {
-		if (descMap.attrInt == 0x0020) {
+	def results = []
+
+	if (descMap.value) {
+		def rawValue = zigbee.convertHexToInt(descMap.value)
+		def batteryValue = null
+
+		if (rawValue == 0xFF) {
+			// Log invalid readings to info for analytics and skip sending an event.
+			// This would be a good thing to watch for and form some sort of device health alert if too many come in.
+			log.info "Invalid battery reading returned"
+		} else if (descMap.attrInt == BATTERY_VOLTAGE_ATTR && !isIkeaDimmer()) { // Ignore from IKEA Dimmer if it sends this since it is probably 0
 			def minVolts = 2.3
 			def maxVolts = 3.0
-			def batteryValueVoltage = zigbee.convertHexToInt(descMap.value) / 10
-			batteryValue = Math.round(((batteryValueVoltage - minVolts) / (maxVolts - minVolts)) * 100)
-		} else if (descMap.attrInt == 0x0021) {
-			batteryValue = zigbee.convertHexToInt(descMap.value) / 2
+			def batteryValueVoltage = rawValue / 10
+
+			batteryValue = Math.round(((batteryValueVoltage- minVolts) / (maxVolts - minVolts)) * 100)
+		} else if (descMap.attrInt == BATTERY_PERCENT_ATTR) {
+			batteryValue = rawValue / 2
 		}
-	}
 
-	if (batteryValue != null) {
-		sendEvent(name: "battery", value: batteryValue)
+		if (batteryValue != null) {
+			results << createEvent(name: "battery", value: batteryValue, unit: "%")
```

Comment:
This was missing before but should add `descriptionText` to make the `battery` event consistent with other DTHs

Reply:
@tpmanley I believe it would be filled in by the capability definition's default. I can certainly set it based on other uses in our platform. Though I would think we'd prefer to stick with the capability definition's when possible? (I need to check to see if they differ.)